File: go/stack/stack.go
/**
 * 这里之所以将 Stack 定义为一个接口，是因为 Stack 的底层物理实现是有多种的
 * 而不是说但有一种实现，因此这里我们将 Stack 定义为一个接口，并定义一个 Stack
 * 接口该有的各种操作方法，至于其底层实现则不加限制
 */
package stack

type Stack interface {
	GetSize() int
	IsEmpty() bool
	Push(elem interface{})
	Pop() (interface{}, error)
	Peek() (interface{}, error)
}

File: js/Stack/Stack.js
const Arr = require('../MyArray/MyArray');

class Stack {
  constructor(capacity = 100) {
    this.data = new Arr(capacity);
  }

  getLength() {
    return this.data.getLength();
  }

  getCapacity() {
    return this.data.getCapacity();
  }

  isEmpty() {
    return this.data.isEmpty();
  }

  push(elem) {
    this.data.addLast(elem);
  }

  pop() {
    return this.data.removeLast();
  }

  peek() {
    return this.data.get(this.getLength() - 1);
  }

  toString() {
    let str = 'Stack:\n';
    str += 'Top: ';
    const length = this.getLength();
    for (let i = length - 1; i > 0 ; i--) {
      str += this.data.get(i) + ', ';
    }

    if (length !== 0) {
      str += this.data.get(0) + ' Bottom';
    }
    return str;
  }
}

module.exports = Stack;